### PR TITLE
Added the ability for users to use config files with V3

### PR DESCRIPTION
**Pull request by EpicLiem**
I added the ability for people to add a config file using --config. There are some conflicts with a few new commits, so I still need to resolve them. Link this pr to issue #981.

### CODE REVIEW

File: src/revChatGPT/V3.py [hunk 1]
```diff
@@ -42,6 +43,40 @@ def __init__(
...
+            if not config.get("top_p") is None:
+                top_p = config.get("top_p")
+            if not config.get("presence_penalty") is None:
+                presence_penalty = config.get("presence_penalty")
+            if not config.get("frequency_penalty") is None:
+                frequency_penalty = config.get("frequency_penalty")
+            if not config.get("reply_count") is None:
+                reply_count = config.get("reply_count")
+            if not config.get("system_prompt") is None:
+                system_prompt = config.get("system_prompt")
```
Comment (Arborym) [inline src/revChatGPT/V3.py:76]: All these should be switched to follow De-Morgan identities:
```py
        if config is not None:
            if config.get("api_key") is not None:
                api_key = config.get("api_key")
            elif api_key is None:
                # equivalent to config.get("api_key") is None and api_key is None
                raise Exception("No api key")
            if config.get("engine") is not None:
                engine = config.get("engine")
            if config.get("proxy") is not None:
                proxy = config.get("proxy")
            if config.get("max_tokens") is not None:
                max_tokens = config.get("max_tokens")
            if config.get("temperature") is not None:
                temperature = config.get("temperature")
            if config.get("top_p") is not None:
                top_p = config.get("top_p")
            if config.get("presence_penalty") is not None:
                presence_penalty = config.get("presence_penalty")
            if config.get("frequency_penalty") is not None:
                frequency_penalty = config.get("frequency_penalty")
            if config.get("reply_count") is not None:
                reply_count = config.get("reply_count")
            if config.get("system_prompt") is not None:
                system_prompt = config.get("system_prompt")```

File: src/revChatGPT/V3.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ class Chatbot:
 
     def __init__(
         self,
-        api_key: str,
+        api_key: str = None,
```
Comment (EpicLiem) [inline src/revChatGPT/V3.py:25]: Api key can be included in config, so we shouldn't require it.

File: src/revChatGPT/V3.py [hunk 3]
```diff
@@ -40,6 +41,40 @@ def __init__(
...
         self.api_key = api_key
         self.proxy = proxy
+
+        self.system_prompt = system_prompt
+        self.max_tokens = max_tokens
+        self.temperature = temperature
+        self.top_p = top_p
+        self.presence_penalty = presence_penalty
+        self.frequency_penalty = frequency_penalty
+        self.reply_count = reply_count
```
Comment (EpicLiem) [inline src/revChatGPT/V3.py:51]: Moved these up, so config will overwrite them.

**Review by Arborym — CHANGES_REQUESTED** [review]
Some formatting things to consider in comments

**Comment by joshuabrink1** [comment]
Wouldn't a more dry solution not require a `config` class variable and rather expand save and load functions, to work with config like:
```python
    def save(self, file: str, *keys: str) -> None:
        """
        Save the Chatbot configuration to a JSON file
        """
        with open(file, "w", encoding="utf-8") as f:
            json.dump(
                {
                    key: self.__dict__[key]
                    for key in self.__get_filtered_class_keys(*keys)
                },
                f,
                indent=2,
            )

    def load(self, file: str, *keys: str) -> None:
        """
        Load the Chatbot configuration from a JSON file
        """
        with open(file, encoding="utf-8") as f:
            loaded_config = json.load(f)

            self.__dict__.update(
                {
                    key: loaded_config[key]
                    for key in self.__get_filtered_class_keys(*keys)
                }
            )

    def __get_filtered_class_keys(self, *keys: str) -> list:
        """
        Get filtered class keys from the given keys. 
        """
        class_keys = self.__dict__.keys() - {"session"}
        if not keys:
            return class_keys

        # Check if all passed keys are valid
        if invalid_keys := set(keys) - {"not"} - class_keys:
            raise ValueError(
                f"Invalid keys: {invalid_keys}",
            )

        # Remove the passed keys from the class keys.
        if keys[0] is "not":
            given_keys = keys[1:]
            return [key for key in class_keys if key not in given_keys]
        else:
            # Only return specified keys that are in class_keys
            return [key for key in keys if key in class_keys]
```

This would allow for commands like: 
- `!load chat.json top_p temperature` to only load `top_p` and `temperature` from a config file
- `!save chat.json not top_p temperature` to save all fields except `top_p` and `temperature` to a config file
- And simplify implementation with cli:
```python
    if config:       
        chatbot = ChatbotCLI(args.api_key)
        try:
            chatbot.load(config)
        except:
            print(f"Error: {args.config} could not be loaded")
            sys.exit()
    else:
        chatbot = ChatbotCLI(
            api_key=args.api_key,
            system_prompt=args.base_prompt,
            proxy=args.proxy,
            temperature=args.temperature,
            top_p=args.top_p,
            reply_count=args.reply_count,
        )
```

**Comment by EpicLiem** [comment]
> Wouldn't a more dry solution not require a `config` class variable and rather expand save and load functions, to work with config like:
> 
> ```python
>     def save(self, file: str, *keys: str) -> None:
>         """
>         Save the Chatbot configuration to a JSON file
>         """
>         with open(file, "w", encoding="utf-8") as f:
>             json.dump(
>                 {
>                     key: self.__dict__[key]
>                     for key in self.__get_filtered_class_keys(*keys)
>                 },
>                 f,
>                 indent=2,
>             )
> 
>     def load(self, file: str, *keys: str) -> None:
>         """
>         Load the Chatbot configuration from a JSON file
>         """
>         with open(file, encoding="utf-8") as f:
>             loaded_config = json.load(f)
> 
>             self.__dict__.update(
>                 {
>                     key: loaded_config[key]
>                     for key in self.__get_filtered_class_keys(*keys)
>                 }
>             )
> 
>     def __get_filtered_class_keys(self, *keys: str) -> list:
>         """
>         Get filtered class keys from the given keys. 
>         """
>         class_keys = self.__dict__.keys() - {"session"}
>         if not keys:
>             return class_keys
> 
>         # Check if all passed keys are valid
>         if invalid_keys := set(keys) - {"not"} - class_keys:
>             raise ValueError(
>                 f"Invalid keys: {invalid_keys}",
>             )
> 
>         # Remove the passed keys from the class keys.
>         if keys[0] is "not":
>             given_keys = keys[1:]
>             return [key for key in class_keys if key not in given_keys]
>         else:
>             # Only return specified keys that are in class_keys
>             return [key for key in keys if key in class_keys]
> ```
> 
> This would allow for commands like:
> 
>     * `!load chat.json top_p temperature` to only load `top_p` and `temperature` from a config file
> 
>     * `!save chat.json not top_p temperature` to save all fields except `top_p` and `temperature` to a config file
> 
>     * And simplify implementation with cli:
> 
> 
> ```python
>     if config:       
>         chatbot = ChatbotCLI(args.api_key)
>         try:
>             chatbot.load(config)
>         except:
>             print(f"Error: {args.config} could not be loaded")
>             sys.exit()
>     else:
>         chatbot = ChatbotCLI(
>             api_key=args.api_key,
>             system_prompt=args.base_prompt,
>             proxy=args.proxy,
>             temperature=args.temperature,
>             top_p=args.top_p,
>             reply_count=args.reply_count,
>         )
> ```

I'll refactor my code to use this approach

**Comment by EpicLiem** [comment]
Does anyone know why the PR isn't updating? I added changes to branch, and it updated, but they're not appearing here.

**Comment by EpicLiem** [comment]
It seems I needed to add another commit, and now it updated the pr.

**Review by Arborym — CHANGES_REQUESTED** [review]

File: src/revChatGPT/V3.py (hunk 2 above)
Comment (acheong08) [reply src/revChatGPT/V3.py:25]: Need checks to see if it's included in at least 1

**Comment by acheong08** [comment]
Conflicts

File: src/revChatGPT/V3.py [hunk 4]
```diff
@@ -245,6 +247,47 @@ def load(self, file: str, *convo_ids: str) -> bool:
...
+                        self.presence_penalty = config.get("presence_penalty")
+                    if config.get("frequency_penalty") is not None:
+                        self.frequency_penalty = config.get("frequency_penalty")
+                    if config.get("reply_count") is not None:
+                        self.reply_count = config.get("reply_count")
+                    if config.get("max_tokens") is not None:
+                        self.max_tokens = config.get("max_tokens")
+                    if config.get("system_prompt") is not None:
+                        self.system_prompt = config.get("system_prompt")
+                        self.reset(system_prompt=self.system_prompt)
```
Comment (acheong08) [inline src/revChatGPT/V3.py:279]: Not necessary. Use `or` to check for `None`

**Review by acheong08 — APPROVED** [review]

**Review by acheong08 — CHANGES_REQUESTED** [review]
.

File: src/revChatGPT/V3.py (hunk 4 above)
Comment (acheong08) [reply src/revChatGPT/V3.py:279]: I'll fix it

**Comment by EpicLiem** [comment]
@acheong08 The custom submit key breaks things if a submit key isn't given because it then tries to reference the key_binding var when it's never defined. Should I fix that in this pr?

**Review by acheong08 — APPROVED** [review]

File: src/revChatGPT/V3.py [hunk 5]
```diff
@@ -409,17 +453,32 @@ def main() -> NoReturn:
...
-        top_p=args.top_p,
-        reply_count=args.reply_count,
-    )
+    # Load config
+    if args.config is not None:
+        # Initialize chatbot
+        chatbot = ChatbotCLI("placeholder")
+        no_api_key = False
+        if args.api_key is None:
+            no_api_key = True
```
Comment (Arborym) [inline src/revChatGPT/V3.py:463]:
```suggestion
        no_api_key = args.api_key is None
```

**Review by acheong08 — APPROVED** [review]

**Pull request merged**